### PR TITLE
解决日志系统未初始化完成时，apollo 的加载日志没法输出问题

### PR DESCRIPTION
**Pull request by klboke**
## What's the purpose of this PR

> 解决问题：

解决 spring boot 方式集成时，配置  `apollo.bootstrap.eagerLoad.enabled` 情况下，apollo 的加载日志没法输出问题。

> 设计原理：

封装一个基于 guava-cache 的日志输出内容存储容器 `DeferredLogUtil`，将 apollo 初始化时的日志信息，添加到延迟输出的日志容器。在日志系统加载完成也就是 EnvironmentPostProcessor 事件立刻回放，最终实现的效果和不开启 `apollo.bootstrap.eagerLoad.enabled` 是一样的，设计过程参考了 [spring-boot#DeferredLog](https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/logging/DeferredLog.java)

### CODE REVIEW

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java [hunk 1]
```diff
@@ -58,8 +59,9 @@ private void initialize() {
       updateConfig(m_configRepository.getConfig(), m_configRepository.getSourceType());
     } catch (Throwable ex) {
       Tracer.logError(ex);
-      logger.warn("Init Apollo Local Config failed - namespace: {}, reason: {}.",
-          m_namespace, ExceptionUtil.getDetailMessage(ex));
+      String logWarnMsg = "Init Apollo Local Config failed - namespace: {}, reason: {}.";
+      logger.warn(logWarnMsg, m_namespace, ExceptionUtil.getDetailMessage(ex));
+      DeferredLogUtil.warn(logger, logWarnMsg, m_namespace, ExceptionUtil.getDetailMessage(ex));
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java:64]: Writing the `DeferredLogUtil.xxx` statements in many places don't seem like a very good way since it adds duplication and developer would probably forget to call this method in the future.

How about we wrap the original logger so that we don't need to duplicate the log statements? e.g.

```java
private static final Logger logger = DeferredLogger.wrap(LoggerFactory.getLogger(DefaultConfig.class));
```

or even

```java
private static final Logger logger = DeferredLoggerFactory.getLogger(DefaultConfig.class));
```

File: apollo-client/src/main/java/com/ctrip/framework/apollo/spring/boot/ApolloApplicationContextInitializer.java [hunk 2]
```diff
@@ -91,7 +92,8 @@ public void initialize(ConfigurableApplicationContext context) {
   protected void initialize(ConfigurableEnvironment environment) {
 
     if (environment.getPropertySources().contains(PropertySourcesConstants.APOLLO_BOOTSTRAP_PROPERTY_SOURCE_NAME)) {
-      //already initialized
+      //already initialized ,  replay the logs that were printed before the logging system was initialized
+      DeferredLogUtil.replayTo();
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/spring/boot/ApolloApplicationContextInitializer.java:96]: In a non-spring-boot environment, this code won't be executed, which means these logs remains in the memory forever?

I think we could do it in a more precise way, e.g.

**1. We only enable DeferredLogger when eager load is enabled**
Since `com.ctrip.framework.apollo.spring.boot.ApolloApplicationContextInitializer#postProcessEnvironment` will be the first lines of code executed, we could set a flag here

**2. In our DeferredLogger implementation, we cache the logs if the flag is true**

**3. We need to find a point to reset the flag to replay the cached logs and stop caching logs**
I assume there would be some sort of events published when the spring logging system is ready but we need to find it out.

Comment (klboke) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/spring/boot/ApolloApplicationContextInitializer.java:96]: - The log will not always be stored in the memory. In an environment where the replay condition cannot be triggered, the log will be recycled after one minute

- In the current client environment, I feel that whether it is the current design or the new delayed log design, it is possible to directly use the apollo.bootstrap.enabled tag to replay. The timing of the replay at this time is also more appropriate

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java (hunk 1 above)
Comment (klboke) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java:64]: This is a good idea

Comment (klboke) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java:64]: After looking at it, it is not very easy to achieve. How about this?
```java
    public static void debug(Logger logger, String message, Object... objects) {
        logger.debug(message,objects);
        add(logger, Level.DEBUG, message, objects, null);
    }
```
Use it like this：
```java
      DeferredLogUtil.error(logger, "Initialize DefaultServerProvider failed.", ex);    
```

**Comment by nobodyiam** [comment]
Here is my 2 cents, what do you think?
1. we keep the original logger interface but with a different implementation, e.g. a composite logger
2. we add a `deferred` flag to control the logger logic
3. in normal cases, the `deferred` flag is always false, so the log behavior is the same as the old one
4. in eager load case, we first set the `deferred` flag to true to cache the logs and after logging system is initialized, we set the flag to false and flush the cached logs

image

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java [hunk 3]
```diff
@@ -59,7 +59,7 @@ private void initialize() {
     } catch (Throwable ex) {
       Tracer.logError(ex);
       logger.warn("Init Apollo Local Config failed - namespace: {}, reason: {}.",
-          m_namespace, ExceptionUtil.getDetailMessage(ex));
+              m_namespace, ExceptionUtil.getDetailMessage(ex));
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java:62]: please use apollo code style to format the code

File: apollo-client/src/main/java/com/ctrip/framework/apollo/spring/boot/ApolloApplicationContextInitializer.java [hunk 4]
```diff
@@ -91,7 +92,8 @@ public void initialize(ConfigurableApplicationContext context) {
   protected void initialize(ConfigurableEnvironment environment) {
 
     if (environment.getPropertySources().contains(PropertySourcesConstants.APOLLO_BOOTSTRAP_PROPERTY_SOURCE_NAME)) {
-      //already initialized
+      //already initialized ,  replay the logs that were printed before the logging system was initialized
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/spring/boot/ApolloApplicationContextInitializer.java:95]:
```suggestion
      //already initialized, replay the logs that were printed before the logging system was initialized
```

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLog.java [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+ * <pre class="code">
+ *  public void debug(String s, Object... objects);
+ *  public void info(String s, Object... objects);
+ *  public void warn(String s, Object... objects);
+ *  public void error(String s, Throwable throwable);
+ * </pre>
+ * @author kl (http://kailing.pub)
+ * @since 2021/5/19
+ */
+public class DeferredLog implements Logger {
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLog.java:17]:
```suggestion
public class DeferredLogger implements Logger {
```

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLog.java [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+    }
+
+    @Override
+    public void warn(String s, Object o) {
+        logger.warn(s, o);
+    }
+
+    @Override
+    public void warn(String s, Object... objects) {
+        logger.warn(s, objects);
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLog.java:229]: I was thinking we add a check here so that we don't print the log twice? e.g.

```java
if (deferred) {
    DeferredLogCache.warn(logger, s, objects);
} else {
    logger.warn(s, objects);
}
```

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+    private static void add(Logger logger, Level level, String message, Object[] objects, Throwable throwable) {
+        if (isEnabled()) {
+            Line logLine = new Line(level, message, objects, throwable, logger);
+            LOG_CACHE.put(LOG_INDEX.incrementAndGet(), logLine);
+        }
+    }
+
+    public static void replayTo() {
+        if (isEnabled()) {
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:66]: I think we need to disable deferred log first? or there might be some logs missing, i.e. those logs printed after the for statement

File: apollo-core/src/test/java/com/ctrip/framework/apollo/core/utils/DeferredLogCacheTest.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author kl (http://kailing.pub)
+ * @since 2021/5/11
+ */
+public class DeferredLogCacheTest {
```
Comment (nobodyiam) [inline apollo-core/src/test/java/com/ctrip/framework/apollo/core/utils/DeferredLogCacheTest.java:12]: This ut is a little bit simple...

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLog.java (hunk 5 above)
Comment (nobodyiam) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLog.java:17]: btw, I think we need to add some unit test for this class, or the code coverage rate will drop a lot..

**Comment by nobodyiam** [comment]
@klboke please fix the failed tests

image

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+/**
+ * Delayed log printing utility class, used only for logging when Apollo is initialized
+ *
+ * @author kl (http://kailing.pub)
+ * @since 2021/5/11
+ */
+public final class DeferredLogCache {
+
+  public static final int MAX_LOG_SIZE = 1000;
+  private static final AtomicBoolean ENABLED = new AtomicBoolean(false);
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:21]: How about we move this ENABLED property to DeferredLogger?

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Delayed log printing utility class, used only for logging when Apollo is initialized
+ *
+ * @author kl (http://kailing.pub)
+ * @since 2021/5/11
+ */
+public final class DeferredLogCache {
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:18]: I think we won't reuse this class elsewhere, how about we make it package-private or even move the whole class as a private inner static class of DeferredLogger?
```suggestion
final class DeferredLogCache {
```

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+  private static final AtomicInteger LOG_INDEX = new AtomicInteger(0);
+  private static final Cache<Integer, Line> LOG_CACHE = CacheBuilder.newBuilder()
+      .maximumSize(MAX_LOG_SIZE)
+      .expireAfterWrite(1, TimeUnit.MINUTES)
+      .build();
+
+  private DeferredLogCache() {
+  }
+
+  public static synchronized void enableDeferredLog() {
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:31]: It looks like the `synchronized` is not necessary?

File: apollo-core/src/test/java/com/ctrip/framework/apollo/core/utils/DeferredLoggerTest.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+  @Test
+  public void testDeferredLog() {
+    DeferredLogCache.enableDeferredLog();
+
+    DeferredLoggerTest.logger.error("errorLogger");
+    DeferredLoggerTest.logger.info("inFoLogger");
+    DeferredLoggerTest.logger.warn("warnLogger");
+    DeferredLoggerTest.logger.debug("debugLogger");
+    DeferredLogCache.replayTo();
```
Comment (nobodyiam) [inline apollo-core/src/test/java/com/ctrip/framework/apollo/core/utils/DeferredLoggerTest.java:60]: I think we need to test the outContent is empty before replayTo method?

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java (hunk 11 above)
Comment (klboke) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:31]: Yes, I forgot to remove it here

File: apollo-core/src/test/java/com/ctrip/framework/apollo/core/utils/DeferredLoggerTest.java (hunk 12 above)
Comment (klboke) [reply apollo-core/src/test/java/com/ctrip/framework/apollo/core/utils/DeferredLoggerTest.java:60]: Yes it is

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java (hunk 9 above)
Comment (klboke) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:21]: It doesn't feel very good, but here you can add a logic that only opens once, what do you think?

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java (hunk 10 above)
Comment (klboke) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:18]: Same as above

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+  public static void disable() {
+    state.set(0);
+  }
+
+  public static boolean isEnabled() {
+    return state.get() == 1;
+  }
+
+  public static void replayTo(){
+    DeferredLogCache.replayTo();
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java:50]: How about we disable the state here so that DeferredLogCache is simply a cache which doesn't need to care whether the deferred logger is enabled or not?
```suggestion
    disable();
    DeferredLogCache.replayTo();
```

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        assert logLine != null;
+        Logger logger = logLine.getLogger();
+        Level level = logLine.getLevel();
+        String message = logLine.getMessage();
+        Object[] objects = logLine.getObjects();
+        Throwable throwable = logLine.getThrowable();
+        logTo(logger, level, message, objects, throwable);
+      }
+      clear();
+    }
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:66]: I think DeferredLogCache could remain simple by not caring whether DeferredLogger is enabled or not.
Actually it could even be renamed to `LogCache` which means it's simple a cache.

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java (hunk 13 above)
Comment (klboke) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java:50]: It makes sense, my thinking has been stuck on `DeferredLogUtils`

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+
+  }
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:68]: I think DeferredLogCache could remain simple by not caring whether DeferredLogger is enabled or not.
Actually it could even be renamed to LogCache which means it's simple a cache.
```suggestion
  public static void replayTo() {
    for (int i = 1; i <= LOG_INDEX.get(); i++) {
      Line logLine = LOG_CACHE.getIfPresent(i);
      assert logLine != null;
      Logger logger = logLine.getLogger();
      Level level = logLine.getLevel();
      String message = logLine.getMessage();
      Object[] objects = logLine.getObjects();
      Throwable throwable = logLine.getThrowable();
      logTo(logger, level, message, objects, throwable);
    }
    clear();
  }
```

Comment (klboke) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:68]: yes,it is

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java [hunk 16]
```diff
@@ -29,7 +29,8 @@
  * @author Jason Song(song_s@ctrip.com)
  */
 public class DefaultConfig extends AbstractConfig implements RepositoryChangeListener {
-  private static final Logger logger = LoggerFactory.getLogger(DefaultConfig.class);
+
+  private static final Logger logger = DeferredLoggerFactory.getLogger(DefaultConfig.class);
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java:33]: I think `com.ctrip.framework.apollo.internals.AbstractConfigFile` also need to use `DeferredLogger`

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java [hunk 17]
```diff
@@ -0,0 +1,419 @@
...
+    return logger.isWarnEnabled();
+  }
+
+  @Override
+  public void warn(String s) {
+    if (isEnabled()) {
+      DeferredLogCache.warn(logger, s);
+    } else {
+      logger.warn(s);
+    }
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java:287]: Is it necessary to implement all the methods? maybe by calling the overloaded methods?

```suggestion
    this.warn(s, (Object) null);
```

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java (hunk 16 above)
Comment (klboke) [reply apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java:33]: yes，It is

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java (hunk 17 above)
Comment (klboke) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java]: Basic coverage common methods

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+
+  public static void debug(Logger logger, String message, Throwable throwable, Object... objects) {
+    add(logger, Level.DEBUG, message, objects, throwable);
+  }
+
+  public static void info(Logger logger, String message, Throwable throwable, Object... objects) {
+    add(logger, Level.INFO, message, objects, throwable);
+  }
+
+  public static void warn(Logger logger, String message, Throwable throwable, Object... objects) {
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:38]: with this signature, we can't make sure throwable and objects are not null at the same time, maybe we could add another overload method?

```java
public static void warn(Logger logger, String message, Object... objects) {}
public static void warn(Logger logger, String message, Throwable throwable) {}
```

Comment (nobodyiam) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:38]: btw, looks like the ci failed, please take a look~

Comment (klboke) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:38]: DeferredLogCache is not open to the public, the current api meets the needs, the compilation error is caused by my negligence

Comment (nobodyiam) [reply apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:54]: OK, then I would suggest adding a method javadoc here to make it clear that user should not pass throwable and objects at the same time in case it was ignored in the future

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java [hunk 19]
```diff
@@ -0,0 +1,465 @@
...
+      DeferredLogCache.debug(logger, s, null, o);
+    } else {
+      logger.debug(s, o);
+    }
+  }
+
+  @Override
+  public void debug(String s, Object o, Object o1) {
+    if (isEnabled()) {
+      DeferredLogCache.debug(logger, s, null, o1);
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java:172]: the object instance `o` is missing

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java [hunk 20]
```diff
@@ -0,0 +1,465 @@
...
+      DeferredLogCache.warn(logger, s, null, o, o1);
+    } else {
+      logger.warn(s, o, o1);
+    }
+  }
+
+  @Override
+  public void warn(String s, Throwable throwable) {
+    if (isEnabled()) {
+      DeferredLogCache.warn(logger, s, null);
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogger.java:350]: the throwable instance is missing

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+      Level level = logLine.getLevel();
+      String message = logLine.getMessage();
+      Object[] objects = logLine.getObjects();
+      Throwable throwable = logLine.getThrowable();
+      logTo(logger, level, message, objects, throwable);
+    }
+    clear();
+  }
+
+  public static void clear() {
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:82]: this method could be package-private
```suggestion
 static void clear() {
```

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+    LOG_CACHE.invalidateAll();
+    LOG_INDEX.set(0);
+  }
+
+  public static long logSize() {
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:87]: this method could be package-private
```suggestion
  static long logSize() {
```

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+    return LOG_CACHE.size();
+  }
+
+  static void logTo(Logger logger, Level level, String message, Object[] objects,
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:91]: this method could be private
```suggestion
  private static void logTo(Logger logger, Level level, String message, Object[] objects,
```

File: apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+        } else {
+          logger.error(message, objects);
+        }
+        break;
+      default:
+        throw new IllegalStateException("Unexpected value: " + level);
+    }
+  }
+
+  static class Line {
```
Comment (nobodyiam) [inline apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java:127]: this class could be private
```suggestion
  privarte static class Line {
```

File: apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java [hunk 25]
```diff
@@ -45,7 +45,8 @@
...
```
Comment (nobodyiam) [inline apollo-client/src/main/java/com/ctrip/framework/apollo/internals/DefaultConfig.java:49]: Please also change LocalFileConfigRepository and RemoteConfigRepository to use `DeferredLogger`, I tested in my local environment and found the actual cause won't be printed in eager load mode

**Review by nobodyiam — APPROVED** [review]
LGTM
